

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Documentation</title>
    <link rel="stylesheet" type="text/css" href="/static/css/styles.css">
    <link href="https://fonts.googleapis.com/css2?family=Montserrat:wght@300;400;700&display=swap" rel="stylesheet">
    <link crossorigin="anonymous" rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.5.3/dist/css/bootstrap.min.css" integrity="sha384-TX8t27EcRE3e/ihU7zmQxVncDAy5uIKz4rEkgIXeMed4M0jlfIDPvg6uqKI2xXr2" />
    <style>
        .navbar {
            background-color: #4F80C0;
        }

        .example-button {
            padding: 10px 20px;
            font-size: 16px;
            cursor: pointer;
            background-color: #4F80C0;
            color: white;
            border-radius: 12px;
            border: none;
            text-decoration: none;
            display: inline-block;
        }
        .example-button:hover {
            background-color: #0056b3;
        }

        p a, p a:hover{
          color: #BA839B;
          text-decoration: underline;
        }

        p a, p a{
          color: #BA839B;
          text-decoration: none;
        }

    </style> 
</head>
<body>
<div class="page">
      <div class="header">
        <nav class="navbar navbar-expand-lg navbar-light text-white">
          <a class="navbar-brand" href="/results/search/">
            <h1 class="text-white text-decoration-none">ProteoCast 
              <span style="font-size: 0.6em; color: white;"> Documentation</span></h1>
          </a>
          <button class="navbar-toggler" type="button" aria-label="Toggle navigation" aria-controls="navbarNav" data-target="#navbarNav" data-toggle="collapse">
            <span class="navbar-toggler-icon"></span>
          </button>
          <div class="collapse navbar-collapse justify-content-between" id="navbarNav">
            <ul class="navbar-nav ml-auto" id="navbarItems">
              <li class="nav-item" id="nav-home">
                <a class="nav-link text-white" href="/results/search/">Home</a>
              </li>
              <li class="nav-item" id="nav-documentation">
                <a class="nav-link text-white" href="/documentation">Documentation</a>
              </li>
              <li class="nav-item" id="nav-download">
                <a class="nav-link text-white" href="/download">Download</a>
              </li>
              {% comment %} <li class="nav-item" id="nav-drosophiladb">
                <a class="nav-link text-white" href="/drosophiladb">DrosophilaDB</a>
              </li> {% endcomment %}
              <li class="nav-item" id="nav-publications">
                <a class="nav-link text-white" href="/publications">Publications</a>
              </li>
              <li class="nav-item" id="nav-publications">
                <a class="nav-link text-white" href="/past_jobs">Past Jobs</a>
              </li>

            </ul>
          </div>
        </nav>
      </header>
      <div class="content">
        <div class="container">
         <br>
        <p>The <b>ProteoCast Server</b> allows you to rapidly and easily explore how and to what extent missense mutations affect protein function. ProteoCast relies on GEMME, a highly efficient, unsupervised, and fully interpretable variant effect predictor (<a href="https://academic.oup.com/mbe/article/36/11/2604/5548199" target="_blank">Laine et al., 2019</a>). By focusing on how protein residues segregate along the topology of evolutionary trees, GEMME has proven instrumental for studying protein stability, function, and disease mechanisms in several studies -- see, for instance, (<a href="https://www.nature.com/articles/s41586-023-06328-6" target="_blank">Tsuboyama et al., 2023</a>). It was extensively tested against over 2.5M experimental measurements (<a href="https://academic.oup.com/gbe/article/15/11/evad201/7344676" target="_blank">Abakarova et al., 2023</a>, <a href="https://academic.oup.com/mbe/article/36/11/2604/5548199" target="_blank">Laine et al., 2019</a>). It is among the top-performing methods on the widely adopted <a href="https://proteingym.org/">ProteinGym benchmark</a>. </p>
          
     <!--   <p>For a quick start, you can see our 
          <a href="https://youtu.be/k44U0WNwzC4">YouTube tutorial</a>:</p>
        <p style="text-align:center">
          <iframe height="315" allow="accelerometer; autoplay; clipboard-write;
      encrypted-media; gyroscope; picture-in-picture" src="https://www.youtube.com/embed/k44U0WNwzC4" title="YouTube video player" width="560" frameborder="0" allowfullscreen=""></iframe>
        </p> -->
        <p> You want to know more, collaborate, or you encounter a problem with the server? Please feel free to 
          <a href="mailto: elodie.laine@sorbonne-universite.fr">contact us</a>!</p>
        <h3> Input </h3>
        <p> Please provide a multiple sequence alignement in FASTA or A3M format. The protein of interest should be the first one in the alignment and should not contain any gap. We recommend generating the alignment with the highly efficient MMseqs2-based protocol implemented in <a href = "https://colab.research.google.com/github/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb"  target="_blank">ColabFold</a> for an optimal balance between speed and accuracy </a>. Using this protocol, the input alignment should contain at least a couple hundred sequences to obtain reliable predictions (<a href="https://academic.oup.com/gbe/article/15/11/evad201/7344676" target="_blank">Abakarova et al., 2023</a>).</p>
        <p> Optionally, you can opt for mapping the results onto a 3D structure. If the query sequence header is a Uniprot identifier (e.g. P07643), then ProteoCast will automatically retrieve the corresponding 3D model from the <a href = "https://alphafold.com/">AlphaFold Database</a>. Alternatively, you can provide a custom single-chain PDB file.</p>
        
        <h3> Output </h3>
        <p>Please check our example 
          <a href="https://proteocast.ijm.fr/results/results/?q=FBpp0301274">result page</a>. </p>

        <h4>Mutational landscape</h4>
        <p> In this interactive heatmap, each 
          <b>square</b> represents a numerical estimate of the impact of a given amino acid substitution at a given position. The darker the color, the more negative GEMME scores, and thus the stronger the predicted effect. 
        <div class="text-center">
          <img src="/static/browser/images/gemme_mutational_landscape_example.png" id='panels' alt="mutLand" style="width: 80%; height: auto;">
        </div>
          By hovering the mouse cursor on the representation, a contextual window will indicate the mutation (here P244H) and its predicted raw score (here -6.84). You may choose to visualise the raw predicted scores or discrete classes, namely impactful, uncertain, and neutral.</p>
            <p> The horizontal bar at the bottom of the matrix reflects the confidence in the predictions. Purple indicates reliable predictions, and white unreliable ones. We consider predictions as unreliable when evolutionary information derived from the input alignment is too scarce.
          </p>
        <h4>Input alignement quality assessment</h4>
        <p>This plot gives an overview of the input alignment, where each horizontal line depicts a sequence and its color indicates its similarity with the query. Gaps correspond to white interruptions. The black curve reports the percentage of sequences that have an amino acid (as opposed to a gap) at each position of the query sequence. This representation is identical to the one used in <a href = "https://colab.research.google.com/github/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb"  target="_blank">ColabFold</a> and we adapted the code from there.</p>
        <div class="text-center">
          <img src="/static/browser/images/msaRepresentation_example.jpg" id='panels' alt="MSA_cov" style="width: 80%; height: auto;">
        </div>
        <h4>Predicted score distributions for variant classification</h4>
        <p> This plot gives an overview of the predicted score distribution, fitted with a mixture of three Gaussians. To minimize biases, it does not account fro low-confidence predictions. At one end of the spectrum, mutations close to  zero extending down to the median of the middle Gaussian are classified as neutral. At the other end, mutations with very negative scores are classified as impactful, until they are more likely to belong to the middle Gaussian than the leftmost one. Mutations falling in between are categorised as uncertain. The vertical lines delineate the three classes. </p>
        <div class="text-center">
          <img src="/static/browser/images/GMM_example.jpg" id='panels' alt="GMM_distrib" style="width: 80%; height: auto;">
        </div>
        
        <h4> Segmented mutational sensitivity profile to identify local patterns and  biases</h4>
        <p> This plot shows the per-residue mutational sensitivity, defined as the average GEMME score over the 19 possible substitutions. The profile is segmented using the <a href="https://CRAN.R-project.org/package=fpopw">FPOP algorithm</a>. FPOP identifies changepoints in the signal through 'functional pruning'. Informally, each detected changepoint signifies a shift in the mean. This representation allows for emphasizing protein segments under stronger or weaker selective pressure than their surrounding background. Yellow: the segment mean is lower than the two neighbouring segments. Purple: the segment mean is higher than the two neighbouring segments. Red: the segment mean is higher than one neighbour and lower than the other one.</p>
  <div class="text-center">
    <img src="/static/browser/images/SegProfile_example.png" id='panels' alt="segmentation profile" style="width: 80%; height: auto;">
  </div>
        <h4> Localisation of mutations  on the 3D structure</h4>
        <p> This interactive 3D model of the protein allows you to explore the localisation of any mutation of interest, whether it is on the surface of the protein or within the core, whether it is located in a secondary structure or a loop region...etc. You can choose between different colour regimes that reflect the original B-factor values (pLDDT for AlphaFold models) or the per-residue sensitivity. For improved clarity, we also provide a binary classification of residues into tolerant or sensitive to mutations. Sensitive residues have more than half of the 19 possible substitutions being impactful. </p>
  <div class="text-center">
    <img src="/static/browser/images/3D_example.jpg" id='panels' alt="3D structure" style="width: 80%; height: auto;">
  </div>
        <h3> Browser compatibility </h3>
        <p> The server's JavaScript code uses 
  
          <a href="https://www.w3schools.com/js/js_es6.asp">ECMAScript 2015 (ES6)</a> 
  features; therefore, it needs 
  Chrome 58+, Edge 15+, Firefox 54+, Safari 10+, or Opera 55+. 
  We have tested this server using the following browsers:</p>
        <div class="tableContainer">
          <table class="table table-bordered">
            <tbody>
              <tr>
                <td>OS</td>
                <td align="center">Version</td>
                <td align="center">Chrome</td>
                <td align="center">Firefox</td>
                <td align="center">Microsoft Edge</td>
                <td align="center">Safari</td>
              </tr>
              <tr>
                <td>Linux</td>
                <td align="center">Ubuntu 22.04</td>
                <td align="center">131.0.6778.139 </td>
                <td align="center">133.0.3</td>
                <td align="center">n/a</td>
                <td align="center">n/a</td>
              </tr>
              <tr>
                <td>MacOS</td>
                <td align="center">Mojave</td>
                <td align="center">87.0.4280.88</td>
                <td align="center">83.0</td>
                <td align="center">n/a</td>
                <td align="center">14.0</td>
              </tr>
              <tr>
                <td>Windows</td>
                <td align="center">Windows 10 Home</td>
                <td align="center">87.0.4280.88</td>
                <td align="center">84.0</td>
                <td align="center">87.0.664.60</td>
                <td align="center">n/a</td>
              </tr>
            </tbody>
          </table>
        </div>
        <p>If you find any problem, please report it to 
          <a href="mailto: marina.abakarova@sorbonne-universite.fr">us</a>.</p>
      </div>


      <footer class="bg-light text-center py-3">
            <p class="mb-0">ProteoCast - This website is free and open to all users and there is no login requirement.</p>
    </footer>
    </div>
  </body></html>
