{% load static %}
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Citation</title>
    <link rel="stylesheet" type="text/css" href="{% static 'css/styles.css' %}">
    <link href="https://fonts.googleapis.com/css2?family=Montserrat:wght@300;400;700&display=swap" rel="stylesheet">
    <link crossorigin="anonymous" rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.5.3/dist/css/bootstrap.min.css" integrity="sha384-TX8t27EcRE3e/ihU7zmQxVncDAy5uIKz4rEkgIXeMed4M0jlfIDPvg6uqKI2xXr2" />
</head>
<body>
    <header class="bg-dark text-white py-3">
        <div class="container">
            <h1 class="mb-0">
                <a href="{% url 'search' %}" class="text-white text-decoration-none">ProteoCast</a>
                <span class="h5"> Documentation</span>
            </h1>
        </div>
    </header>


    <div class="container">
        <p>The 
          <b>ProteoCast Server</b> allows you to rapidly and easily explore how missense mutations can functionaly affect proteins. ProteoCast primary source of information are the protein
  sequence alignements used by
          GEMME (<a href="https://www.ensembl.org/index.htmlhttps://academic.oup.com/mbe/article/36/11/2604/5548199">Laine et al., 2019</a>) running in the  
          pipeline described in (<a href="https://github.com/PhyloSofS-Team/thoraxehttps://academic.oup.com/gbe/article/15/11/evad201/7344676">Abakarova et al., 2023</a>).
          
     <!--   <p>For a quick start, you can see our 
          <a href="https://youtu.be/k44U0WNwzC4">YouTube tutorial</a>:</p>
        <p style="text-align:center">
          <iframe height="315" allow="accelerometer; autoplay; clipboard-write;
      encrypted-media; gyroscope; picture-in-picture" src="https://www.youtube.com/embed/k44U0WNwzC4" title="YouTube video player" width="560" frameborder="0" allowfullscreen=""></iframe>
        </p> -->
        <p> You want to know more, collaborate, or you encounter a problem with the server? Please feel free to 
          <a href="mailto: elodie.laine@sorbonne-universite.fr">contact us</a>!</p>
        <h3> Input </h3>
        <p> Please provide your Multiple Sequence Alignement using <a href = "https://colab.research.google.com/github/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb">CollabFold</a>. 
          Optionally, you can provide <i>Uniprot ID</i> (e.g. P07643) or upload a PDB file for mapping ProteoCast predictions onto the 3D model of the protein.
  Ideally, 100 or more sequences should be used for the MSA to allow relevant predictions (<a href="https://academic.oup.com/gbe/article/15/11/evad201/7344676">Abakarova et al., 2023</a>).</p>
        
        <h3> Output </h3>
        <p>Please check our example 
          <a href="https://proteocast.ijm.fr/results/results/?q=FBpp0301274">result page</a>. </p>
        <p> Ases provides a comprehensive visualization of the results with pre-tuned parameters (nodes and edges colors and
  shapes) and with clear connections between the different representations of the biological objects (graph,
  forest of trees, multiple sequence alignments, sequences, genes and tables with ids and annotations).</p>

        <h4>GEMME mutational landscape</h4>
        <p> This interactive graph summarizes the transcript variability observed for the query gene across the selected
  species. Each 
          <b>square</b> corresponds to a single amino acid substitution and its predicted impact score is color coded. 
        <div class="text-center">
          <img src="{% static 'browser/images/gemme_mutational_landscape_example.png' %}" id='panels' alt="mutLand" style="width: 80%; height: auto;">
        </div>
          By hovering the mouse cursor on the representation, a contextual window will indicate the precise mutation (here P244H) and
          its calculated score (here -6.84).
  <p></p>
        <h4>Raw Predicted Score distributions</h4>
        <p> To systematically and automatically evaluate the agreement between our evolutionary-based mutational effect estimates 
          and the known annotations, we classified all variants into three categories: <span style="color: red">impactful</span>, <span style="color: rgb(209, 126, 140)">uncertain</span> or <span style="color: rgb(18, 176, 224)">neutral</span>. We achieved 
          this by fitting a mixture of three Gaussians to GEMME predicted score distribution using the expectation-maximization 
          algorithm. This Gaussian Mixture Model (GMM) was used to tailor the classification of each protein’s individual mutational landscape. To minimize biases, we discarded low-confidence predictions.
  </p>
        <div class="text-center">
          <img src="{% static 'browser/images/GMM_example.jpg' %}" id='panels' alt="GMM_distrib" style="width: 80%; height: auto;">
        </div>
        <h4>Multiple Sequences Alignement (MSA) quality assessment</h4>
        <p>This panel will show you the MSA coverage.</p>
        <div class="text-center">
          <img src="{% static 'browser/images/msaRepresentation_example.jpg' %}" id='panels' alt="GMM_distrib" style="width: 80%; height: auto;">
        </div>
        <h4> Protein Segmentation profile for identifying local mutational impact biases</h4>
        <p> This plot shows the moving average GEMME score along the protein with automatically identified segments characterized by
          local high GEMME average. Within these segments, the functional impact of missense mutations is <u>on average</u> higher than 
          in the surrounding regions.
        </p>
  <div class="text-center">
    <img src="{% static 'browser/images/SegProfile_example.png' %}" id='panels' alt="segmentation profile" style="width: 80%; height: auto;">
  </div>
        <h4> Localisation of mutations  on the 3D structure</h4>
        <p> This interactive 3D model of the protein allows you to explore the localisation of your chosen mutation in its conformational context. </p>
  <div class="text-center">
    <img src="{% static 'browser/images/3D_example.jpg' %}" id='panels' alt="3D structure" style="width: 80%; height: auto;">
  </div>
        <h3> Browser compatibility </h3>
        <p> The server's JavaScript code uses 
  
          <a href="https://www.w3schools.com/js/js_es6.asp">ECMAScript 2015 (ES6)</a> 
  features; therefore, it needs 
  Chrome 58+, Edge 15+, Firefox 54+, Safari 10+, or Opera 55+. 
  We have tested this server using the following browsers:</p>
        <div class="tableContainer">
          <table class="table table-bordered">
            <tbody>
              <tr>
                <td>OS</td>
                <td align="center">Version</td>
                <td align="center">Chrome</td>
                <td align="center">Firefox</td>
                <td align="center">Microsoft Edge</td>
                <td align="center">Safari</td>
              </tr>
              <tr>
                <td>Linux</td>
                <td align="center">Ubuntu 22.04</td>
                <td align="center">131.0.6778.139 </td>
                <td align="center">133.0.3</td>
                <td align="center">n/a</td>
                <td align="center">n/a</td>
              </tr>
              <tr>
                <td>MacOS</td>
                <td align="center">Mojave</td>
                <td align="center">87.0.4280.88</td>
                <td align="center">83.0</td>
                <td align="center">n/a</td>
                <td align="center">14.0</td>
              </tr>
              <tr>
                <td>Windows</td>
                <td align="center">Windows 10 Home</td>
                <td align="center">87.0.4280.88</td>
                <td align="center">84.0</td>
                <td align="center">87.0.664.60</td>
                <td align="center">n/a</td>
              </tr>
            </tbody>
          </table>
        </div>
        <p>If you find any problem, please report it to 
          <a href="mailto: marina.abakarova@sorbonne-universite.fr">us</a>.</p>
      </div>


    <footer class="bg-light text-center py-3">
        <div class="container">
            <p class="mb-0">ProteoCast - This website is free and open to all users and there is no login requirement.</p>
        </div>
    </footer>
    <script src="https://code.jquery.com/jquery-3.5.1.slim.min.js" integrity="sha384-DfXdz2htPH0lsSSs5nCTpuj/zy4C+OGpamoFVy38MVBnE+IbbVYUew+OrCXaRkfj" crossorigin="anonymous"></script>
