<!DOCTYPE html>
<html lang="en">
<head>

<!-- Google Tag Manager -->
<script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
  new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
  j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
  'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
  })(window,document,'script','dataLayer','GTM-NPCKRJ3W');</script>
  <!-- End Google Tag Manager -->
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Documentation</title>
    <link rel="stylesheet" type="text/css" href="/static/css/styles.css">
    <link href="https://fonts.googleapis.com/css2?family=Montserrat:wght@300;400;700&display=swap" rel="stylesheet">
    <link crossorigin="anonymous" rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.5.3/dist/css/bootstrap.min.css" integrity="sha384-TX8t27EcRE3e/ihU7zmQxVncDAy5uIKz4rEkgIXeMed4M0jlfIDPvg6uqKI2xXr2" />
    <!-- Bootstrap CSS -->
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.6.2/dist/css/bootstrap.min.css">
    <!-- jQuery -->
    <script src="https://code.jquery.com/jquery-3.6.4.min.js"></script>
    <!-- Popper.js -->
    <script src="https://cdn.jsdelivr.net/npm/@popperjs/core@2.11.8/dist/umd/popper.min.js"></script>
    <!-- Bootstrap JS -->
    <script src="https://cdn.jsdelivr.net/npm/bootstrap@4.6.2/dist/js/bootstrap.min.js"></script>
    <style>
        .navbar {
            background-color: #4F80C0;
        }

        .example-button {
            padding: 10px 20px;
            font-size: 16px;
            cursor: pointer;
            background-color: #4F80C0;
            color: white;
            border-radius: 12px;
            border: none;
            text-decoration: none;
            display: inline-block;
        }
        .example-button:hover {
            background-color: #0056b3;
        }
        p a, p a:hover{
          color: #BA839B;
          text-decoration: underline;
        }

        p a, p a{
          color: #BA839B;
          text-decoration: none;
        }

        ul a, ul a:hover{
          color: #BA839B;
          text-decoration: underline;
        }

        ul a, ul a{
          color: #BA839B;
          text-decoration: none;
        }
        .justify-text {
          text-align: justify;
        }

    </style> 
</head>
<body>
  <!-- Google Tag Manager (noscript) -->
<noscript><iframe src="https://www.googletagmanager.com/ns.html?id=GTM-NT9G797M"
  height="0" width="0" style="display:none;visibility:hidden"></iframe></noscript>
  <!-- End Google Tag Manager (noscript) -->
<div class="page">
  <div class="header">
    <nav class="navbar navbar-expand-lg navbar-light text-white">
      <a class="navbar-brand" href="/results/search/">
        <h1 class="text-white text-decoration-none">ProteoCast</h1>
      </a>
      <button 
        class="navbar-toggler" 
        type="button" 
        aria-label="Toggle navigation" 
        data-toggle="collapse" 
        data-target="#navbarNav" 
        aria-controls="navbarNav" 
        aria-expanded="false" 
        aria-haspopup="true">
        <span class="navbar-toggler-icon"></span>
      </button>
      <div class="collapse navbar-collapse justify-content-between" id="navbarNav">
        <ul class="navbar-nav ml-auto" id="navbarItems">
          <li class="nav-item" id="nav-home">
            <a class="nav-link text-white" href="/results/search/">Home</a>
          </li>
          <li class="nav-item" id="nav-documentation">
            <a class="nav-link text-white" href="/documentation">Documentation</a>
          </li>
          <li class="nav-item" id="nav-download">
            <a class="nav-link text-white" href="/download">Download</a>
          </li>
          <li class="nav-item" id="nav-past-jobs">
            <a class="nav-link text-white" href="/past_jobs">Past Jobs</a>
          </li>
          <li class="nav-item" id="nav-publications">
            <a class="nav-link text-white" href="/publications">Publications</a>
          </li>
        </ul>
      </div>
    </nav>
  </div>
      </header>
      <div class="content">
        <div class="container">
         <br>
        <p class="justify-text">The <b>ProteoCast Server</b> allows you to rapidly and easily explore how and to what extent missense mutations affect protein function. ProteoCast relies on GEMME, a highly efficient, unsupervised, and fully interpretable variant effect predictor (<a href="https://academic.oup.com/mbe/article/36/11/2604/5548199" target="_blank">Laine et al., 2019</a>). By focusing on how protein residues segregate along the topology of evolutionary trees, GEMME has proven instrumental for studying protein stability, function, and disease mechanisms in several studies -- see, for instance, <a href="https://www.nature.com/articles/s41586-023-06328-6" target="_blank">Tsuboyama et al., 2023</a>. It was extensively tested against over 2.5M experimental measurements (<a href="https://academic.oup.com/gbe/article/15/11/evad201/7344676" target="_blank">Abakarova et al., 2023</a>, <a href="https://academic.oup.com/mbe/article/36/11/2604/5548199" target="_blank">Laine et al., 2019</a>). It is among the top-performing methods on the widely adopted <a href="https://proteingym.org/" target="_blank">ProteinGym benchmark</a>. </p>
          
     <!--   <p>For a quick start, you can see our 
          <a href="https://youtu.be/k44U0WNwzC4">YouTube tutorial</a>:</p>
        <p style="text-align:center">
          <iframe height="315" allow="accelerometer; autoplay; clipboard-write;
      encrypted-media; gyroscope; picture-in-picture" src="https://www.youtube.com/embed/k44U0WNwzC4" title="YouTube video player" width="560" frameborder="0" allowfullscreen=""></iframe>
        </p> -->
        <h5> <b>You want to know more, collaborate, or you encounter a problem with the server? Please feel free to 
              <a href="mailto: elodie.laine@sorbonne-universite.fr" class="text-danger">contact us</b></a>!</h5>  

              <h3 class="py-2"> Input </h3>
        <p class="justify-text"> Please provide a multiple sequence alignement in FASTA or A3M format. The protein of interest should be the first one in the alignment and should not contain any gap. We recommend generating the alignment with the highly efficient MMseqs2-based protocol implemented in <a href = "https://colab.research.google.com/github/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb"  target="_blank">ColabFold</a> for an optimal balance between speed and accuracy</a>. Using this protocol, the input alignment should contain at least a couple hundred sequences to obtain reliable predictions (<a href="https://academic.oup.com/gbe/article/15/11/evad201/7344676" target="_blank">Abakarova et al., 2023</a>).</p>
        <p class="justify-text" > Optionally, you can opt for mapping the results onto a 3D structure. The most straightforward way to use this functionality is detailed below:
          <ul>
            <li class="justify-text"> The query sequence corresponds exactly to a Uniprot entry. You can indicate the Uniprot identifier and ProteoCast will automatically retrieve the corresponding 3D model from the <a href = "https://alphafold.com/" target="_blank">AlphaFold Database</a>.
            </li>
            <li class="justify-text"> The query sequence do not match a Uniprot entry. You may generate a 3D model along with the alignment using <a href = "https://colab.research.google.com/github/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb"  target="_blank">ColabFold</a> and upload it directly. </li>
          </ul> 
          <p class="justify-text">Alternatively, you may provide an experimental structure from the <a href="https://www.ebi.ac.uk/pdbe/" target='_blank'>Protein Data Bank</a>, or a custom 3D model. In that case, please make sure that the sequence is exactly the same as the query. Any user-defined 3D structure or model should be in PDB format and contain only one protein chain.</p>
        
        <h3> Output </h3>
        <p>Please check our example 
          <a href="https://proteocast.ijm.fr/results/results/?q=job20250108135938" target="_blank">result page</a>. </p>

        <h4>Mutational landscape</h4>
        <p class="justify-text"> The predicted mutational landscape is displayed as two interactive heatmaps, where each <b>square</b> corresponds to a given amino acid substitution at a given position. The heatmap dimension is thus 20 by the length of the query protein sequence. The heatmap <b>RAW SCORES</b> contains the numerical estimates predicted by GEMME. The darker the color, the more negative the score, and thus the stronger the predicted effect. The heatmap <b>VARIANT CLASSES</b> indicates whether each mutation is considered as neutral (blue), uncertain (pink) or impactful (red) by ProteoCast. For pre-computed results on the <i>Drosophila melanogaster</i>, we optionally provide a third heatmap <b>SNPs</b> with the raw scores in grey tones and known SNPs highlighted in colors, blue for population polymorphisms and red for lethal mutations.  </p>
        <div class="text-center">
         <img src="/static/browser/images/snapshot1_mutland.png" id='panels' class="my-4" alt="mutLand" style="width: 80%; height: auto;">
        </div>
          <p class="justify-text">By hovering the mouse cursor on the representation, a contextual window will specify the mutation and give some information about it (predicted raw score, class...). You may zoom in a particular region of interest.</p>
          <p class="justify-text" > The horizontal bar at the bottom of the heatmap reflects the <b>confidence</b> in the predictions. Dark blue indicates reliable predictions, and white unreliable ones. We consider predictions as unreliable when evolutionary information derived from the input alignment is too scarce.
          </p>
        <h4>Input alignement quality assessment</h4>
        <p class="justify-text">This plot gives an overview of the input alignment, where each horizontal line depicts a sequence and its color indicates its similarity with the query. Gaps correspond to white interruptions. The black curve reports the percentage of sequences that have an amino acid (as opposed to a gap) at each position of the query sequence. This representation is identical to the one used in <a href = "https://colab.research.google.com/github/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb"  target="_blank">ColabFold</a> and we adapted the code from there.</p>
        <div class="text-center">
          <img src="https://proteocast.ijm.fr/data/15806/3.FBpp0288697_msaRepresentation.jpg" id='panels' class="my-4" alt="MSA_cov" style="width: 80%; height: auto;">
        </div>
        <h4>Predicted score distributions for variant classification</h4>
        <p class="justify-text"> This plot gives an overview of the predicted score distribution, fitted with a mixture of three Gaussians. To minimize biases, low-confidence predictions are excluded. At one end of the spectrum, mutations close to zero extending down to the median of the middle Gaussian are classified as <b>neutral</b>. At the other end, mutations with very negative scores are classified as <b>impactful</b>, until they are more likely to belong to the middle Gaussian than the leftmost one. Mutations falling in between are categorised as <b>uncertain</b>. The two vertical lines indicate the boundaries of the three classes. </p>
        <div class="text-center">
          <img src="/static/browser/images/snapshot2_GMM.png" id='panels' class="my-4" alt="GMM_distrib" style="width: 80%; height: auto;">
        </div>
        
        <h4> Segmented mutational sensitivity profile</h4>
        <p class="justify-text"> This interactive plot shows the per-residue mutational sensitivity, defined as the average GEMME score over the 19 possible substitutions. The values are scaled between 0 and 1. Residues highly sensitive to mutations will have a value close to 1, and residues highly tolerant a value close to 0. The profile is segmented using the <a href="https://CRAN.R-project.org/package=fpopw" target="_blank">FPOP algorithm</a>, which identifies changepoints in the signal through 'functional pruning'. Informally, each detected changepoint signifies a shift in the mean. This representation allows for emphasizing protein segments under stronger or weaker selective pressure than their surrounding background. <b>Purple</b>: the segment mean is higher than the two neighbouring segments. <b>Red</b>: the segment mean is higher than one neighbour and lower than the other one. A pLDDT track is added on top of the mutational sensitivity profile to improve interpretability -- for instance, toward the identification of putative binding and regulatory sites in unstructured regions. The four classes for pLDDT are those defined in the AlphaFold Database, orange: very low, yellow: low, light blue: medium, dark blue: high.  

         </p>
  <div class="text-center">
    <img src="/static/browser/images/snapshot3_segments.png" id='panels' class="my-4" alt="segmentation profile" style="width: 80%; height: auto;">
  </div>
        <h4> Mapping of predictions on the 3D structure</h4>
        <p class="justify-text"> This interactive <a href="https://github.com/molstar/pdbe-molstar" target='_blank'>Molstar plugin</a> allows you to explore the localisation of any mutation of interest, whether it is on the surface of the protein or within the core, whether it is part of a well-define secondary structure or an unstructured  loop...etc. You can choose between different colour regimes that reflect the original <b>B-factor values (pLDDT for AlphaFold models)</b> or the <b>per-residue mutational sensitivity</b>. For improved clarity, we also provide a binary classification of residues into <b>tolerant</b> or <b>sensitive</b> to mutations. Sensitive residues have more than half of the 19 possible substitutions being impactful. On each representation, you may visualise the segments located in unstructured regions (very low or low pLDDT) and whose sensitivity stands out against their surroundings. They are colored in purple or red, following the color scheme used for the segmented profile above. They are displayed using a cartoon-like representation in which the protein backbone is shown like a clay or <b>putty</b> structure. The size of the tube is proportional to the property being examined (pLDDT, mutational sensitivity or residue class). On the figure below, we can clearly see that the segment containing S146 has higher mutational sensitivity than its surroundings. </p>
  <div class="text-center">
    <img src="/static/browser/images/snapshot4_structure.png" id='panels' class="my-4" alt="3D structure" style="width: 80%; height: auto;">
  </div>
        <h3> Browser compatibility </h3>
        <p> The server's JavaScript code uses 
  
          <a href="https://www.w3schools.com/js/js_es6.asp" target="_blank" >ECMAScript 2015 (ES6)</a> 
  features; therefore, it needs 
  Chrome 58+, Edge 15+, Firefox 54+, Safari 10+, or Opera 55+. 
  We have tested this server using the following browsers:</p>
        <div class="tableContainer">
          <table class="table table-bordered">
            <tbody>
              <tr>
                <td>OS</td>
                <td align="center">Version</td>
                <td align="center">Chrome</td>
                <td align="center">Firefox</td>
                <td align="center">Microsoft Edge</td>
                <td align="center">Safari</td>
              </tr>
              <tr>
                <td>Linux</td>
                <td align="center">Ubuntu 22.04</td>
                <td align="center">131.0.6778.139 </td>
                <td align="center">133.0.3</td>
                <td align="center">n/a</td>
                <td align="center">n/a</td>
              </tr>
              <tr>
                <td>MacOS</td>
                <td align="center">Mojave</td>
                <td align="center">87.0.4280.88</td>
                <td align="center">83.0</td>
                <td align="center">n/a</td>
                <td align="center">14.0</td>
              </tr>
              <tr>
                <td>Windows</td>
                <td align="center">Windows 10 Home</td>
                <td align="center">87.0.4280.88</td>
                <td align="center">84.0</td>
                <td align="center">87.0.664.60</td>
                <td align="center">n/a</td>
              </tr>
            </tbody>
          </table>
        </div>

      </div>


      <footer class="bg-light text-center py-3">
        <p class="mb-0">ProteoCast - This website is free and open to all users and there is no login requirement.</p>
</footer>
<script>
    document.addEventListener('DOMContentLoaded', function () {
      const footer = document.querySelector('footer');
      footer.style.display = 'none'; // Initially hide the footer

      window.addEventListener('scroll', function () {
        if ((window.innerHeight + window.scrollY) >= document.body.offsetHeight) {
          footer.style.display = 'block'; // Show the footer when scrolled to the bottom
        } else {
          footer.style.display = 'none'; // Hide the footer otherwise
        }
      });
    });
    </script>
    </div>
  </body></html>
