

<!DOCTYPE html>
<html lang="en">
<!-- Google Tag Manager -->
<script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
  new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
  j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
  'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
  })(window,document,'script','dataLayer','GTM-NPCKRJ3W');</script>
  <!-- End Google Tag Manager -->
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>DrosophilaDB</title>
    <link rel="stylesheet" type="text/css" href="/static/css/styles.css">
    <link href="https://fonts.googleapis.com/css2?family=Montserrat:wght@300;400;700&display=swap" rel="stylesheet">
    <link crossorigin="anonymous" rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.5.3/dist/css/bootstrap.min.css" integrity="sha384-TX8t27EcRE3e/ihU7zmQxVncDAy5uIKz4rEkgIXeMed4M0jlfIDPvg6uqKI2xXr2" />
    <style>
        .navbar {
            background-color: #4F80C0;
        }

        .example-button {
            padding: 10px 20px;
            font-size: 16px;
            cursor: pointer;
            background-color: #4F80C0;
            color: white;
            border-radius: 12px;
            border: none;
            text-decoration: none;
            display: inline-block;
        }
        .example-button:hover {
            background-color: #0056b3;
        }

        p a, p a:hover{
          color: #BA839B;
          text-decoration: underline;
        }

        p a, p a{
          color: #BA839B;
          text-decoration: none;
        }
        ul a, ul a:hover{
          color: #BA839B;
          text-decoration: underline;
        }

        ul a, ul a{
          color: #BA839B;
          text-decoration: none;
        }
    </style> 
</head>
<body>
<div class="page">
      <div class="header">
        <nav class="navbar navbar-expand-lg navbar-light text-white">
          <a class="navbar-brand" href="/results/search/">
            <h1 class="text-white text-decoration-none">ProteoCast
              <span style="font-size: 0.6em; color: white;"> for <em>Drosophila melanogaster</em></span></h1>
          </a>
          <button class="navbar-toggler" type="button" aria-label="Toggle navigation" aria-controls="navbarNav" data-target="#navbarNav" data-toggle="collapse">
            <span class="navbar-toggler-icon"></span>
          </button>
          <div class="collapse navbar-collapse justify-content-between" id="navbarNav">
            <ul class="navbar-nav ml-auto" id="navbarItems">
            <li>
            <form action="/results/results/" method="get" class="search-form">
            <input type="text" name="q" placeholder="e.g., yki-PG or FBpp0288697" value="">
            <button type="submit">Search</button>
          </form>
        </li>
      </ul>
          </div>
        </nav>
      </header>
      <div class="content">
        <div class="container">
         <br>
         <p> Through this portal, you can explore and interactively visualise <b>ProteoCast</b> predictions for <em>Drosophila melanogaster</em> proteome, including all protein isoforms (proteoforms) resulting from alternative splicing.</p>
        <p> Try it out with your favorite protein! Please provide <a href="https://flybase.org/" target="_blank">FlyBase</a> protein ID (FBpp) or proteoform symbol (i.e. yki-PF) in the top-right search dialog.</p>
        
        <p> For more information and resources, please refer to the <a href="/documentation">documentation</a>.</p>
        <h4>How it works: </h4> 
        <p>
        <ul>
            <li>
                ProteoCast performed <b>proteome-wide predictions</b> of all possible amino acid variants  using <a href = "http://www.lcqb.upmc.fr/GEMME/Home.html">GEMME</a> (<a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC6805226/">Laine et al., 2019</a>). We generated the input alignments for GEMME using the MMseqs2-based protocol implemented in <a href = "https://colab.research.google.com/github/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb">ColabFold</a> (<a href = "https://www.nature.com/articles/s41592-022-01488-1">Mirdita et al., 2022</a>, <a href = "https://www.nature.com/articles/nbt.3988">Steinegger and SÃ¶ding, 2017</a>).
            </li>
            <br>
            <li>
                ProteoCast classified the variants into <b>impactful, uncertain or neutral</b> by fitting a Gaussian Mixture Model to GEMME raw score distribution for each protein. Any residue with more than 10 (out of 19) impactful substitutions is considered as sensitive to mutations. 
            </li>
            <br>
            <li>
                ProteoCast applied the <b><a href="https://CRAN.R-project.org/package=fpopw">FPOP</a> segmentation algorithm</b> (<a href="https://link.springer.com/article/10.1007/s11222-016-9636-3">Maidstone et al., 2017</a>) on GEMME sensitivity profile (per-residue average scores), taking AlphaFold pLDDT into account. Segments that stand out from their surrounding background represent potential binding sites and regulatory motifs.  
            </li>
            <br>
            <li>
                We <b>validated of our proteom-wide predictions</b> using different approaches. 
                <br>First, by assessing the occurence of all missense mutations amongst two different resources for genetic variation in natural populations. The <i>Drosophila melanogaster</i> Genetic Reference Panel (<a href = "https://www.nature.com/articles/nature10811">Mackay et al., 2012</a>) gives us access to natural polymorphism fixed in 200 inbred lines while the Drosophila Evolution over Space and Time (<a href="https://academic.oup.com/mbe/article/38/12/5782/6361628">Kapun et al., 2021</a>; <a href="https://www.biorxiv.org/content/10.1101/2024.11.10.622744v1.abstract">Nunez et al. , 2024</a>) gives us access to natural polymorphism occuring in the wild (sampling period 2009 - 2021). Second, by confronting our predictions to the <a href="https://flybase.org/">FlyBase</a> dataset of developmentally lethal mutations and referenced hypomorphs. Finally, we use existing databases for protein Post Translational Modifications and Short Linear Motifs to interpret residue mutational sensitivity.
            </li>
        </ul></p>

        <div class="text-center">
          <img src="/static/browser/images/ProteoCast_overview.jpg" id='panels' alt="mutLand" style="width: 100%; height: auto;">
        </div>

      </div>


           <footer class="bg-light text-center py-3">
            <p class="mb-0">ProteoCast - This website is free and open to all users and there is no login requirement.</p>
      </footer>
    </div>
  </body></html>
