{% load static %}
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>DrosophilaDB</title>
    <link rel="stylesheet" type="text/css" href="{% static 'browser/styles.css' %}">
    <link href="https://fonts.googleapis.com/css2?family=Montserrat:wght@300;400;700&display=swap" rel="stylesheet">

    <style>
        body {
            font-family: Montserrat, sans-serif;
            line-height: 1;
            margin: 0;
            padding: 0;
            display: flex;
            flex-direction: column;
            align-items: center;
            min-height: 100vh;
        }
        .main-content {
            flex-grow: 1;
            width: 100%;
        }
        .results-container {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin: 20px;
            gap: 20px;
            width: 100%;
        }
        footer {
            width: 100%;
            text-align: center;
            padding: 10px;
            background-color: #f1f1f1;
            position: fixed;
            bottom: 0;
            color: grey;
        }
            /* Center the image */
        .image-container {
            display: flex;           /* Enables flexbox */
            justify-content: center; /* Centers horizontally */
            align-items: center;     /* Centers vertically */
            height: 100vh;           /* Full viewport height */
            margin: 0;               /* Remove default margin */
        }

        /* Adjust image size relative to screen */
        img {
            max-width: 120%;  /* Maximum width: 90% of the viewport width */
            max-height: 120%; /* Maximum height: 90% of the viewport height */
            object-fit: contain; /* Ensures the image maintains aspect ratio */
        }
    </style>
</head>
<body>
    <header>
        <div class="header-content">
            <h1 style="font-size: 2em; text-align: left;">
                <a href="{% url 'search' %}" style="text-decoration: none;">ProteoCast</a> 
                <span style="font-size: 0.8em; color: white;">for <em>Drosophila melanogaster</em></span>
            </h1>
            <form action="{% url 'results' %}" method="get" class="search-form">
                <input type="text" name="q" placeholder="Enter FBpp ID" value="{{ query }}">
                <button type="submit">Search</button>
            </form>
        </div>
    </header>

    <div class="content" style="color: black;">
        <p> Here we provide <b>ProteoCast</b> predictions and visualization for <em>Drosophila melanogaster</em> proteome, including all protein isoforms (proteoforms) resulting from alternative splicing.</p>
        <p> Try it out with your favorite protein. Provide <a href="https://flybase.org/" target="_blank">FlyBase</a> protein ID (FBpp) or proteoform symbol (i.e. yki-PA) in the top-right search dialog.</p>
        
        <p> For more information and resources, please refer to the <a href="{% url 'documentation' %}">documentation</a>.</p>
        <p> <b>How it works: </b></p>
        <ul>
            <li>
                <u>Proteome-wide predictions</u> are made by using the <a href = "https://www.nature.com/articles/s41592-022-01488-1">CollabFold (<a href = "https://www.nature.com/articles/s41592-022-01488-1">Mirdita et al., 2022</a>) implementation of <a href="https://colab.research.google.com/github/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb" target="_blank">MMseqs2</a> (<a href = "https://www.nature.com/articles/nbt.3988">Steinegger and SÃ¶ding, 2017</a>)
                to generate Multiple Sequences Alignements (MSA). These MSA are then used as an input to <a href = "http://www.lcqb.upmc.fr/GEMME/Home.html">GEMME</a> (<a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC6805226/">Laine et al., 2019</a>). 
                GEMME outputs Mutational Landscapes for each proteoform. 
            </li>
            <br>
            <li>
                <u>Classification of our proteom-wide predictions</u> using Gaussian Mixture Models into missense mutations potentially <span style="color: red">impactful</span>, <span style="color: rgb(209, 126, 140)">uncertain</span> or <span style="color: rgb(18, 176, 224)">neutral</span>.
            </li>
            <br>
            <li>
                <u>Validation of our proteom-wide predictions</u> using different approaches. 
                <br>First, by assessing the occurence of all missense mutations amongst two different resources for genetic variation in natural populations. The <i>Drosophila melanogaster</i> Genetic Reference Panel (<a href = "https://www.nature.com/articles/nature10811">Mackay et al., 2012</a>) gives us access to natural polymorphism fixed in 200 inbred lines while the Drosophila Evolution over Space and Time (<a href="https://academic.oup.com/mbe/article/38/12/5782/6361628">Kapun et al., 2021</a>; <a href="https://www.biorxiv.org/content/10.1101/2024.11.10.622744v1.abstract">Nunez et al. , 2024</a>) gives us access to natural polymorphism occuring in the wild (sampling period 2009 - 2021).
                <br>Second, by confronting our predictions to the <a href="https://flybase.org/">FlyBase</a> dataset of developmentally lethal mutations and referenced hypomorphs.
                <br>Finally, we use existing databases for protein Post Translational Modifications and Short Linear Motifs to interpret the residues predicted as <span style="color: rgb(233, 174, 64)">sensitive</span> or <span style="color: rgb(46, 205, 233)">tolerant</span>.
            </li>
        </ul>
        <div class="image-text-container">
        <img src="{% static 'browser/images/ProteoCast_overview.jpg' %}" alt="ProteoCast Logo" class="centered-image" style="width: 50%; height: auto;">
        </div>
    </div>
    <p> You can download the whole archive from <a href="https://doi.org/10.5281/zenodo.10995110"><img src="https://zenodo.org/badge/DOI/10.5281/zenodo.10995110.svg" alt="DOI"></a> (approx. 4GB)</p>
    <footer>
        <p>ProteoCast - This website is free and open to all users and there is no login requirement.</p>
    </footer>
</body>
</html>
