{% load static %}

<!DOCTYPE html>
<html lang="en">
<head>
      <!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-QV6NV55CPK"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-QV6NV55CPK');
</script>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Results</title>
    <!-- CSS personalizadas -->
    <link rel="stylesheet" type="text/css" href="/static/css/styles.css">
    <!-- Plotly -->
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
    <!-- Favicon -->
    <link rel="icon" type="image/x-icon" href="{% static 'favicon.ico' %}">
    <link href="https://fonts.googleapis.com/css2?family=Montserrat:wght@300;400;700&display=swap" rel="stylesheet">
    <link crossorigin="anonymous" rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.5.3/dist/css/bootstrap.min.css" integrity="sha384-TX8t27EcRE3e/ihU7zmQxVncDAy5uIKz4rEkgIXeMed4M0jlfIDPvg6uqKI2xXr2" />
    <!-- Molstar CSS & JS -->
    <link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/pdbe-molstar@latest/build/pdbe-molstar-light.css">
    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/pdbe-molstar@latest/build/pdbe-molstar-plugin.js"></script>
    <style>
        .navbar {
            background-color: #4F80C0;
        }

        .example-button {
            padding: 10px 20px;
            font-size: 16px;
            cursor: pointer;
            background-color: #4F80C0;
            color: white;
            border-radius: 12px;
            border: none;
            text-decoration: none;
            display: inline-block;
        }
        .example-button:hover {
            background-color: #0056b3;
        }

        p a, p a:hover{
          color: #BA839B;
          text-decoration: underline;
        }

        p a, p a{
          color: #BA839B;
          text-decoration: none;
        }

        .heatmap { 
            width: 100%;
            margin: 10px 0; 
        }
        .image-container {
            display: flex;
            justify-content: center; 
            margin: 20px auto; 
        }

        .download-button button {
            padding: 10px 20px;
            font-size: 16px;
            cursor: pointer;
            background-color: #BA839B;
            color: white;
            border-radius: 12px;
            border: none;
            transition: background-color 0.3s ease;
        }
        .download-button button:hover {
            background-color: #0056b3;
        }
        .download-button img {
            width: 30px;
            height: 30px;
            vertical-align: middle;
            margin-right: 10px;
        }

    </style> 
</head>
<body>
    <!-- Google Tag Manager (noscript) -->
<noscript><iframe src="https://www.googletagmanager.com/ns.html?id=GTM-NT9G797M"
    height="0" width="0" style="display:none;visibility:hidden"></iframe></noscript>
    <!-- End Google Tag Manager (noscript) -->
<div class="page">
      <div class="header">
        <nav class="navbar navbar-expand-lg navbar-light text-white">
          <a class="navbar-brand" href="{% url 'search' %}">
            <h1 class="text-white text-decoration-none">ProteoCast 
              <span style="font-size: 0.6em; color: white;"> for {{ prot_name }}</span></h1>
          </a>
          <button 
              class="navbar-toggler" 
              type="button" 
              aria-label="Toggle navigation" 
              data-toggle="collapse" 
              data-target="#navbarNav" 
              aria-controls="navbarNav" 
              aria-expanded="false" 
              aria-haspopup="true">
              <span class="navbar-toggler-icon"></span>
            </button>
          <div class="collapse navbar-collapse justify-content-between" id="navbarNav">
            <ul class="navbar-nav ml-auto" id="navbarItems">
              <li class="nav-item" id="nav-home">
                <a href="{% url 'download_folder' fbpp_id=query %}" class="download-button">
                <button>
                   <img src="{% static 'browser/images/download_icon.png' %}" alt="Download Icon">
                    Download Results
                </button>
               </a>  
             </li>
            </ul>
          </div>
        </nav>
      </div>
      </header>


      <div class="content">
        <div class="container-fluid" style="width:95%;">

        <div class="form-row py-4">
            <h5>Here are ProteoCast results for {{ prot_name }}. {% if query|slice:":4" == "2025" %}Your job ID is {{ query }}.{% endif %} If you need help in exploring and interpreting them, please go to our <a href="/documentation" target="_blank" rel="noopener noreferrer">documentation</a>.</h5>
            {% if warning_message %}
                <p style="color: orange;">
                    <span style="color: black; font-weight: bold;">&#x26A0;</span> {{ warning_message }}
                </p>
            {% endif %}
        </div>

        <div class="form-row" style="display: flex; align-items: flex-start;">
            <div class="heatmap" style="flex: 1; autosize: true;">
                <h2>Mutational Landscape</h2>
                <p> The heatmap gives the predicted effects of all 19-times-the-protein-length possible amino acid substitutions. You can switch between raw scores (the darker the more negative) and variant classification into impactful (red), uncertain (pink) and neutral (blue). The horizontal bar below gives binary confidence scores, blue for reliable predictions, white otherwise.</p>
                {% if query|slice:":4" != "2025" %}
                <div class="buttons" style="margin-left: 20px;">
                    <label style="display: inline-block; margin-right: 20px;">
                        <input type="radio" name="heatmapToggle" value="heatmap_html" checked>
                        RAW SCORES
                    </label>
                    <label style="display: inline-block; margin-right: 20px;">
                        <input type="radio" name="heatmapToggle" value="heatmapClasses_html">
                        VARIANT CLASSES
                    </label>
                    <label style="display: inline-block;">
                        <input type="radio" name="heatmapToggle" value="heatmapSNPs_html">
                        SNPs
                    </label>
                </div>
                {% else %}
                <div class="buttons" style="margin-left: 20px;">
                    <label style="display: inline-block; margin-right: 20px;">
                        <input type="radio" name="heatmapToggle" value="heatmap_html" checked>
                        RAW SCORES
                    </label>
                    <label style="display: inline-block;">
                        <input type="radio" name="heatmapToggle" value="heatmapClasses_html">
                        VARIANT CLASSES
                    </label>
                </div>
                {% endif %}
                <div style="max-height: 600px; overflow-y: hidden; overflow-x: auto; border: 1px solid #ddd; padding: 10px;">
                    {% if heatmap_html %}
                    <div id="heatmap_html" style="display: block;">
                        {{ heatmap_html|safe }}
                    </div>
                    {% endif %}
                    {% if heatmapClasses_html %}
                    <div id="heatmapClasses_html" style="display: none;">
                        {{ heatmapClasses_html|safe }}
                    </div>
                    {% endif %}
                    {% if heatmapSNPs_html %}
                    <div id="heatmapSNPs_html" style="display: none;">
                        {{ heatmapSNPs_html|safe }}
                    </div>
                    {% endif %}
                    {% if not heatmap_html and not heatmapClasses_html and not heatmapSNPs_html %}
                    <p>No results found.</p>
                    {% endif %}
                </div>
            </div>
          </div>

        <script>
            document.addEventListener('DOMContentLoaded', function() {
                // JavaScript for toggling heatmaps
                document.querySelectorAll('input[name="heatmapToggle"]').forEach(button => {
                    button.addEventListener('change', function() {
                        // Get all heatmap divs
                        const heatmapHtmlDiv = document.getElementById('heatmap_html');
                        const heatmapClassesHtmlDiv = document.getElementById('heatmapClasses_html');
                        const heatmapSNPsHtmlDiv = document.getElementById('heatmapSNPs_html');
                        
                        // Toggle visibility based on selected button
                        if (this.value === 'heatmap_html') {
                            heatmapHtmlDiv.style.display = 'block';
                            heatmapClassesHtmlDiv.style.display = 'none';
                            heatmapSNPsHtmlDiv.style.display = 'none';
                        } else if (this.value === 'heatmapClasses_html') {
                            heatmapHtmlDiv.style.display = 'none';
                            heatmapClassesHtmlDiv.style.display = 'block';
                            heatmapSNPsHtmlDiv.style.display = 'none';
                        } else if (this.value === 'heatmapSNPs_html') {
                            heatmapHtmlDiv.style.display = 'none';
                            heatmapClassesHtmlDiv.style.display = 'none';
                            heatmapSNPsHtmlDiv.style.display = 'block';
                        }
                    });
                });
            });
        </script>

        <div class="form-row">
            <div class="form-group col px-4">
                <h2>Input Alignment</h2>
            </div>
            <div class="form-group col px-4">
                <h2>Raw Score Distribution</h2>
            </div>
        </div>
        <div class="form-row">
            <div class="form-group col px-4">
                <p> Each horizontal line in the plot represents a sequence and its color indicates its percentage of identity to the query (on top). The black curve indicates coverage. </p>
            </div>
            <div class="form-group col px-4">
                <p> The dark purple and light pink vertical lines mark the impactful-uncertain and uncertain-neutral thresholds. {% if query|slice:":4" != "2025" %}Known missense mutations are depicted as red dots (lethal) or blue squares or triangles (DGRP or DEST2).{% endif %}</p>
            </div>
        </div>

        <div class="form-row">
            <div class="form-group col px-4">
                {% if fig_msarep %}
                <div class="image-container">
                     <img src="{{ fig_msarep }}" alt="MSA Representation" class="img-fluid">
                </div>
                {% endif %}
            </div>
            <div class="form-group col px-4">
                {% if image_url_1 %}
                <div class="image-container">
                    <img src="{{ image_url_1 }}" alt="GMM" class="img-fluid">
                </div>
                {% endif %}
            </div>
        </div>


        <div class="form-row" style="flex: 1; autosize: true;"> 
            <h2>Sensitivity profile</h2>
            <p> 
                {% if query|slice:":4" == "2025" %} 
                    The mutational sensitivity profile (per-residue GEMME average score) is shown here.
                {% else %}
                    The pLDDT of the AlphaFold-predicted protein 3D model is shown on top and the mutational sensitivity profile (per-residue GEMME average score) at the bottom.
                {% endif %}  Each delineated segment can be considered as displaying a homogeneous mean sensitivity. The color of each segment indicates whether its mean is higher than its two neighbors (purple), lower than its two neighbors (yellow), or higher than one and lower than the other (red).
            </p>
            <div style="max-height: 600px; overflow-y: hidden; overflow-x: auto; border: 1px solid #ddd; padding: 5px;">
            {% if fig_segmentation %}
            <div class="image-container">
                {{ fig_segmentation|safe }}
                <!--<img src="{{ fig_segmentation }}" alt="Segmentation" class="img-fluid" style="width: 80%; display: block; margin: 0 auto;">-->
            </div>
            {% endif %}
            </div>
        </div>

        <!-- 3D structure -->
        <div class="form-row">
            <h2>3D Structure</h2>
            <p> If a UniProt code or a structure is provided, it will be visualized with its B-factors. If the sequence matches the one in the MSA, it can be visualized with GEMME Sensitivity (per-residue GEMME average score) or Residue Class. </p>
        </div>
        <div class="form-row" style="min-height: 600px;">
            <div class="form-group col-10">
                <div id="myViewer" class="viewer-container"></div>
            </div>
            <div class="form-group col-2 pl-5 py-4">
                <div class="buttons">
                    <label>
                        <input type="radio" name="colorToggle" value="bfactor" checked>
                        Color by B-Factors
                    </label>
                    <label>
                        <input type="radio" name="colorToggle" value="gemme">
                        GEMME Sensitivity
                    </label>
                    <label>
                        <input type="radio" name="colorToggle" value="residue_class">
                        Residue Class
                    </label>
                </div>
                <div class="custom-control custom-switch mt-2 py-1">
                    <input type="checkbox" id="segments" name="segments" class="custom-control-input" >
                     <label for="segments" class="custom-control-label text-dark">Highlight Segments</label>
                </div>
            </div>
        </div>
        
        <script>
            // Initialize the viewer
            const viewerInstance = new PDBeMolstarPlugin();
            const viewerContainer = document.getElementById('myViewer');
        
            // Define the PDB URLs for each condition
            const pdbUrls = {
                bfactor: "{{ pdb_url_1 }}",
                gemme: "{{ pdb_url_2 }}",
                residue_class: "{{ pdb_url_3 }}"
            };
        
            // Check if the primary PDB URL is empty
            if (pdbUrls.bfactor === '') {
                console.log("No PDB URL provided for bfactor. Rendering an empty Mol* viewer.");
                viewerInstance.render(viewerContainer, { 
                    bgColor: { r: 255, g: 255, b: 255 },
                    hideCanvasControls: true,
                    layoutIsExpanded: false,
                    hideControls: true
                });
            } else {
                // Define the segments that are 1 (red) or 2 (purple)
                var selectSections = {{ select_segments | default:"[]"|safe }};
                console.log("Segments to highlight:", selectSections);
                console.log("PDB URLs:", pdbUrls.bfactors);
                // Options for the Mol* Viewer
                var options = {
                    customData: { url: '', format: 'pdb' }, // URL will be dynamically updated
                    alphafoldView: true,
                    bgColor: { r: 255, g: 255, b: 255 },
                    hideCanvasControls: [
                        'selection', 'animation', 'controlToggle', 'controlInfo', 
                        'controlZoom', 'controlRotate', 'controlTranslate', 
                        'controlReset', 'controlFullscreen'
                    ],
                    layoutIsExpanded: false,
                    hideControls: true, // Hide the toolbar
                    visualStyle: {
                        polymer: {type: 'cartoon', size: 'uniform'},
                    },
                };
        
                // Function to load the structure and apply coloring
                async function initialiseView(condition = 'bfactor') {
                    console.log("Loading structure for condition:", condition);
                    console.log("PDB URL:", pdbUrls[condition]);
                    options.customData.url = pdbUrls[condition]; // Set the PDB URL
                    options.visualStyle.polymer.color = 'plddt-confidence';
                    await viewerInstance.render(viewerContainer, options);
                    console.log("Viewer instance initialized");
                    viewerInstance.plugin.canvas3d?.setProps({ camera: { manualReset: true } });
                }
        
                async function addStructure(condition) {
                    viewerInstance.plugin.clear();
                    const _data = await viewerInstance.plugin.builders.data.download({ url: pdbUrls[condition] }, { state: { isGhost: false } });
                    const trajectory = await viewerInstance.plugin.builders.structure.parseTrajectory(_data, 'pdb');
                    const model = await viewerInstance.plugin.builders.structure.createModel(trajectory);
                    const structure = await viewerInstance.plugin.builders.structure.createStructure(model);
                    const all = await viewerInstance.plugin.builders.structure.tryCreateComponentStatic(structure, 'all');
                    if (all) {
                        if (condition === 'bfactor') {
                            await viewerInstance.plugin.builders.structure.representation.addRepresentation(all, { type: 'cartoon', color: 'plddt-confidence'});
                        } else {
                            await viewerInstance.plugin.builders.structure.representation.addRepresentation(all, { type: 'cartoon', color: 'uncertainty'});
                        }
                    }
                }
        
                async function loadStructureAndColor(condition) {
                    console.log("Loading structure for condition:", condition);
                    console.log("PDB URL:", pdbUrls[condition]);
                    addStructure(condition);
                }
        
                var checkbox = document.querySelector('input[name="segments"]');
                checkbox.addEventListener('change', function() {
                    if (this.checked) {
                        console.log("Checkbox is checked..");
                        viewerInstance.visual.select({data: selectSections});
                    } else {
                        console.log("Checkbox is not checked..");
                        viewerInstance.visual.clearSelection();
                    }
                });
        
                // Add event listeners to the radio buttons
                document.querySelectorAll('input[name="colorToggle"]').forEach(button => {
                    button.addEventListener('change', function () {
                        checkbox.checked = false;
                        loadStructureAndColor(this.value);
                    });
                });
        
                // Initial load (default to B-Factors)
                initialiseView();
            }
        </script>
       
    </div> 

    <footer class="bg-light text-center py-3">
            <p class="mb-0">ProteoCast - This website is free and open to all users and there is no login requirement.</p>
    </footer>
    <script>
        document.addEventListener('DOMContentLoaded', function () {
          const footer = document.querySelector('footer');
          footer.style.display = 'none'; // Initially hide the footer
    
          window.addEventListener('scroll', function () {
            if ((window.innerHeight + window.scrollY) >= document.body.offsetHeight) {
              footer.style.display = 'block'; // Show the footer when scrolled to the bottom
            } else {
              footer.style.display = 'none'; // Hide the footer otherwise
            }
          });
        });
    </script>

    </div>
  </body></html>


  
